core/chains/evm/txmgr/nonce_tracker.go

```diff
@@ -0,0 +1,201 @@
+package txmgr
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/jpillora/backoff"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+type NonceTrackerTxStore interface {
+	GetAllTransactions(ctx context.Context, fromAddress common.Address, chainID *big.Int) (txs []Tx, err error)
+}
+
+type NonceTrackerClient interface {
+	ConfiguredChainID() *big.Int
+	PendingNonceAt(context.Context, common.Address) (evmtypes.Nonce, error)
+}
+
+type nonceTracker struct {
+	lggr             logger.SugaredLogger
+	nextSequenceMap  map[common.Address]evmtypes.Nonce
+	txStore          NonceTrackerTxStore
+	chainID          *big.Int
+	client           NonceTrackerClient
+	enabledAddresses []common.Address
+
+	sequenceLock sync.RWMutex
+}
+
+func NewNonceTracker(lggr logger.Logger, txStore NonceTrackerTxStore, client NonceTrackerClient) *nonceTracker {
+	lggr = logger.Named(lggr, "NonceTracker")
+	return &nonceTracker{
+		lggr:    logger.Sugared(lggr),
+		txStore: txStore,
+		chainID: client.ConfiguredChainID(),
+		client:  client,
+	}
+}
+
+func (s *nonceTracker) LoadNextSequences(ctx context.Context, addresses []common.Address) {
+	s.sequenceLock.Lock()
+	defer s.sequenceLock.Unlock()
+
+	s.enabledAddresses = addresses
+
+	s.nextSequenceMap = make(map[common.Address]evmtypes.Nonce)
+	for _, address := range addresses {
+		seq, err := s.getSequenceForAddr(ctx, address)
+		if err == nil {
+			s.nextSequenceMap[address] = seq
+		}
+	}
+}
+
+func (s *nonceTracker) getSequenceForAddr(ctx context.Context, address common.Address) (evmtypes.Nonce, error) {
+	seq, err := s.getSequenceFromStore(ctx, address)
+	if err == nil {
+		return seq, nil
+	}
+	s.lggr.Infof("Loading sequence from on-chain for addr: %s", address.String())
+	// Look for nonce on-chain if no tx found for address in TxStore or if error occurred
+	// Returns the nonce that should be used for the next transaction so no need to increment
+	seq, err = s.client.PendingNonceAt(ctx, address)
+	if err == nil {
+		s.lggr.Infof("Setting sequence (%d) from on-chain for addr: %s", seq.Int64(), address.String())
+		return seq, nil
+	}
+	s.lggr.Criticalw("failed to retrieve next sequence from on-chain for address: ", "address", address.String())
+	return seq, err
+}
+
+func (s *nonceTracker) getSequenceFromStore(ctx context.Context, address common.Address) (seq evmtypes.Nonce, err error) {
```

Comment:
Why unpack `FindLatestSequence`'s logic here and not keep it as it is? `GetAllTransactions` seems like a heavier query with no added benefits.

Reply:
I don't think it's a good practice to load every tx from the DB and handle the logic in memory. This is significantly heavier than what we do now. If we decide moving to a generic tx store makes sense then we can make the change, but for now I would suggest to keep it as it is.

Reply:
I agree with Dimitris here. Let's keep it the same. When we refactor the txstore, we can deal with this problem then. For now, let's just focus on the nonce and not change any behavior related to the txstore.